src/base/static/components/molecules/form-field-types/map-drawing-toolbar.js

```diff
@@ -0,0 +1,394 @@
+import classNames from "classnames";
+import React, { Component } from "react";
+import { Map } from "immutable";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+import { translate } from "react-i18next";
+import ColorPicker from "rc-color-picker";
+import "rc-color-picker/assets/index.css";
+
+import {
+  activeMarkerSelector,
+  setMarkers,
+  setActiveMarkerIndex,
+  visibleDrawingToolsSelector,
+  markerPanelVisibilitySelector,
+  activeDrawingToolSelector,
+  setActiveDrawingTool,
+  activeColorpickerSelector,
+  setActiveColorpicker,
+  geometryStyleSelector,
+  setGeometryStyle,
+} from "../../../state/ducks/map-drawing-toolbar";
+
+import { ToolbarButton } from "../../atoms/buttons";
+import { Paragraph } from "../../atoms/typography";
+
+import emitter from "../../../utils/emitter";
+import constants from "../../../constants";
+
+import "./map-drawing-toolbar.scss";
+
+const deleteToolLabels = {
+  "create-marker": "deleteMarkerToolLabel",
+  "create-polyline": "deletePolylineToolLabel",
+  "create-polygon": "deletePolygonToolLabel",
+};
+
+class MapDrawingToolbar extends Component {
+  componentDidMount() {
+    this.props.setMarkers(this.props.markers);
+    emitter.addListener(constants.DRAW_UPDATE_GEOMETRY_EVENT, geometry => {
+      this.props.onChange(this.props.name, geometry);
+    });
+
+    if (this.props.existingGeometry) {
+      emitter.emit(
+        constants.DRAW_INIT_GEOMETRY_EVENT,
+        this.props.existingGeometry.toJS(),
+      );
+      switch (
+        this.props.existingGeometry.get(constants.GEOMETRY_TYPE_PROPERTY_NAME)
+      ) {
+        case "Point":
+          this.props.setActiveDrawingTool(constants.DRAW_CREATE_MARKER_TOOL);
+          this.props.setMarkers(this.props.markers);
+          this.props.setActiveMarkerIndex(
+            this.props.markers.indexOf(
+              this.props.existingGeometryStyle.toJS()[
+                constants.MARKER_ICON_PROPERTY_NAME
+              ],
+            ),
+          );
+          break;
+        case "LineString":
+          this.props.setActiveDrawingTool(constants.DRAW_CREATE_POLYLINE_TOOL);
+          this.props.setGeometryStyle(this.props.existingGeometryStyle.toJS());
+          break;
+        case "Polygon":
+          this.props.setActiveDrawingTool(constants.DRAW_CREATE_POLYGON_TOOL);
+          this.props.setGeometryStyle(this.props.existingGeometryStyle.toJS());
+          break;
+      }
+
+      emitter.emit("place-collection:hide-place", {
+        collectionId: this.props.existingCollectionId,
+        modelId: this.props.existingModelId,
+      });
+    }
+  }
+
+  render() {
+    return (
```

Comment:
There's a decent amount of repetition in all these `ToolbarButton` components. I think they could probably be generated programatically by iterating over some kind of data structure instead. Lmk if you think the readability would benefit from this.

Reply:
It's fine for now - I think a data structure will likely add more complexity than it's worth.

When we move to CSS-in-JS, a lot of the "classNames" verbosity should be removed as well.